File: benchmarks/heap.go
package main

import (
	"container/heap"
	"fmt"
	"os"
)

func hash(x uint64) uint64 {
	x ^= x >> 30
	x *= 0xbf58476d1ce4e5b9
	x ^= x >> 27
	x *= 0x94d049bb133111eb
	x ^= x >> 31
	return x
}

// MaxHeap is the default container/heap usage: a slice with the five-method interface.
// Less reports h[i] > h[j] so Pop yields the largest element.
type MaxHeap []uint64

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i] > h[j] }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MaxHeap) Push(x any) { *h = append(*h, x.(uint64)) }

func (h *MaxHeap) Pop() any {
	old := *h
	n := len(old)
	v := old[n-1]
	*h = old[:n-1]
	return v
}

func main() {
	mode := "sort"
	if len(os.Args) > 1 {
		mode = os.Args[1]
	}
	const N uint64 = 10_000_000
	const K = 1024

	switch mode {
	case "sort":
		h := make(MaxHeap, 0, N)
		for i := uint64(0); i < N; i++ {
			heap.Push(&h, hash(i))
		}
		var acc uint64
		for h.Len() > 0 {
			acc = acc*1000003 + heap.Pop(&h).(uint64)
		}
		fmt.Printf("sort: %d\n", acc)
	case "churn":
		const M = N * 3
		h := make(MaxHeap, 0, K+1)
		var acc uint64
		for i := uint64(0); i < M; i++ {
			heap.Push(&h, hash(i))
			if h.Len() > K {
				acc = acc*1000003 + heap.Pop(&h).(uint64)
			}
		}
		for h.Len() > 0 {
			acc = acc*1000003 + heap.Pop(&h).(uint64)
		}
		fmt.Printf("churn: %d\n", acc)
	case "pushpop":
		const M = N * 3
		h := make(MaxHeap, 0, K)
		for i := uint64(0); i < K; i++ {
			heap.Push(&h, hash(i))
		}
		var acc uint64
		for i := uint64(0); i < M; i++ {
			top := heap.Pop(&h).(uint64)
			acc = acc*1000003 + top
			heap.Push(&h, hash(K+i))
		}
		for h.Len() > 0 {
			acc = acc*1000003 + heap.Pop(&h).(uint64)
		}
		fmt.Printf("pushpop: %d\n", acc)
	default:
		fmt.Fprintf(os.Stderr, "Unknown mode: %s\n", mode)
		os.Exit(1)
	}
}

File: benchmarks/heap.rs
use std::collections::BinaryHeap;
use std::env;

fn hash(mut x: u64) -> u64 {
    x ^= x >> 30;
    x = x.wrapping_mul(0xbf58476d1ce4e5b9);
    x ^= x >> 27;
    x = x.wrapping_mul(0x94d049bb133111eb);
    x ^= x >> 31;
    x
}

fn n() -> u64 {
    // Read N opaquely so the optimizer cannot precompute the whole run.
    env::var("HEAP_N").ok().and_then(|s| s.parse().ok()).unwrap_or(10_000_000)
}

fn main() {
    let mode = env::args().nth(1).unwrap_or_else(|| "sort".to_string());
    let big_n = n();
    const K: u64 = 1024;

    match mode.as_str() {
        "sort" => {
            // BinaryHeap is a max-heap; pop() yields the largest element.
            let mut h: BinaryHeap<u64> = BinaryHeap::with_capacity(big_n as usize);
            for i in 0..big_n {
                h.push(hash(i));
            }
            let mut acc: u64 = 0;
            while let Some(v) = h.pop() {
                acc = acc.wrapping_mul(1000003).wrapping_add(v);
            }
            println!("sort: {}", acc);
        }
        "churn" => {
            let m = big_n * 3;
            let mut h: BinaryHeap<u64> = BinaryHeap::with_capacity((K + 1) as usize);
            let mut acc: u64 = 0;
            for i in 0..m {
                h.push(hash(i));
                if h.len() as u64 > K {
                    acc = acc.wrapping_mul(1000003).wrapping_add(h.pop().unwrap());
                }
            }
            while let Some(v) = h.pop() {
                acc = acc.wrapping_mul(1000003).wrapping_add(v);
            }
            println!("churn: {}", acc);
        }
        "pushpop" => {
            let m = big_n * 3;
            let mut h: BinaryHeap<u64> = BinaryHeap::with_capacity(K as usize);
            for i in 0..K {
                h.push(hash(i));
            }
            let mut acc: u64 = 0;
            for i in 0..m {
                let top = h.pop().unwrap();
                acc = acc.wrapping_mul(1000003).wrapping_add(top);
                h.push(hash(K + i));
            }
            while let Some(v) = h.pop() {
                acc = acc.wrapping_mul(1000003).wrapping_add(v);
            }
            println!("pushpop: {}", acc);
        }
        other => {
            eprintln!("Unknown mode: {}", other);
            std::process::exit(1);
        }
    }
}
